In [44]:
import pandas as pd
import matplotlib.pyplot as plt

# Dataset

Load the dataset

In [45]:
data_df = pd.read_csv("data_district_crime.csv")

In [46]:
data_df.head()

CMPLNT_NUM  ADDR_PCT_CD        BORO_NM CMPLNT_FR_DT CMPLNT_FR_TM  \
0   831526991           67       BROOKLYN   03/31/2018     23:30:00   
1   496736340          120  STATEN ISLAND   03/31/2018     23:20:00   
2   399593692          105         QUEENS   03/31/2018     23:15:00   
3   282376455          110         QUEENS   03/31/2018     23:10:00   
4   201598299           78       BROOKLYN   03/31/2018     23:10:00   

  CMPLNT_TO_DT CMPLNT_TO_TM CRM_ATPT_CPTD_CD         HADEVELOPT  HOUSING_PSA  \
0   03/31/2018     23:37:00        COMPLETED                NaN          NaN   
1   03/31/2018     23:30:00        COMPLETED   RICHMOND TERRACE      10143.0   
2   03/31/2018     23:20:00        COMPLETED                NaN          NaN   
3   03/31/2018     23:20:00        COMPLETED                NaN          NaN   
4   03/31/2018     23:19:00        COMPLETED  572 WARREN STREET       2379.0   

   ...  SUSP_SEX TRANSIT_DISTRICT  VIC_AGE_GROUP        VIC_RACE VIC_SEX  \
0  ...       NaN              NaN        UNKNOWN         UNKNOWN       E   
1  ...         M              NaN            <18           BLACK       F   
2  ...         M              NaN          18-24           BLACK       F   
3  ...         M              NaN          25-44  WHITE HISPANIC       M   
4  ...         F              NaN          25-44           BLACK       F   

  X_COORD_CD Y_COORD_CD   Latitude  Longitude                        Lat_Lon  
0  1003227.0   177460.0  40.653751 -73.931609  (40.653751263, -73.931609227)  
1   962748.0   174174.0  40.644726 -74.077483  (40.644726131, -74.077483159)  
2  1056932.0   200034.0  40.715434 -73.737816  (40.715433534, -73.737816163)  
3  1014990.0   210503.0  40.744414 -73.889065  (40.744413503, -73.889065158)  
4   989141.0   187733.0  40.681967 -73.982367  (40.681967344, -73.982367054)  

[5 rows x 35 columns]

In [47]:
data_df.columns

Index(['CMPLNT_NUM', 'ADDR_PCT_CD', 'BORO_NM', 'CMPLNT_FR_DT', 'CMPLNT_FR_TM',
       'CMPLNT_TO_DT', 'CMPLNT_TO_TM', 'CRM_ATPT_CPTD_CD', 'HADEVELOPT',
       'HOUSING_PSA', 'JURISDICTION_CODE', 'JURIS_DESC', 'KY_CD', 'LAW_CAT_CD',
       'LOC_OF_OCCUR_DESC', 'OFNS_DESC', 'PARKS_NM', 'PATROL_BORO', 'PD_CD',
       'PD_DESC', 'PREM_TYP_DESC', 'RPT_DT', 'STATION_NAME', 'SUSP_AGE_GROUP',
       'SUSP_RACE', 'SUSP_SEX', 'TRANSIT_DISTRICT', 'VIC_AGE_GROUP',
       'VIC_RACE', 'VIC_SEX', 'X_COORD_CD', 'Y_COORD_CD', 'Latitude',
       'Longitude', 'Lat_Lon'],
      dtype='object')

Check for NULL values

In [48]:
data_df.isna().sum()

CMPLNT_NUM                0
ADDR_PCT_CD               0
BORO_NM                  58
CMPLNT_FR_DT              0
CMPLNT_FR_TM              0
CMPLNT_TO_DT          16016
CMPLNT_TO_TM          15966
CRM_ATPT_CPTD_CD          0
HADEVELOPT           103721
HOUSING_PSA          100562
JURISDICTION_CODE        57
JURIS_DESC                0
KY_CD                     0
LAW_CAT_CD                0
LOC_OF_OCCUR_DESC     19392
OFNS_DESC                 4
PARKS_NM             109082
PATROL_BORO              57
PD_CD                    57
PD_DESC                  57
PREM_TYP_DESC           433
RPT_DT                    0
STATION_NAME         106411
SUSP_AGE_GROUP        28133
SUSP_RACE             28124
SUSP_SEX              28124
TRANSIT_DISTRICT     106411
VIC_AGE_GROUP             0
VIC_RACE                  0
VIC_SEX                   0
X_COORD_CD                2
Y_COORD_CD                2
Latitude                  2
Longitude                 2
Lat_Lon                   2
dtype: int64

# Data Cleaning

Cleaning the dataset by dopping certain rows with NA values. Out of many records since only 58 of Borough name are missing, dropping 58 rows. Similar step to few other columns

In [226]:
data_df = data_df.dropna(subset = ['BORO_NM'])
data_df.isna().sum()

CMPLNT_NUM                0
ADDR_PCT_CD               0
BORO_NM                   0
CMPLNT_FR_DT              0
CMPLNT_FR_TM              0
CMPLNT_TO_DT          15957
CMPLNT_TO_TM          15907
CRM_ATPT_CPTD_CD          0
HADEVELOPT           103664
HOUSING_PSA          100499
JURISDICTION_CODE         0
JURIS_DESC                0
KY_CD                     0
LAW_CAT_CD                0
LOC_OF_OCCUR_DESC     19391
OFNS_DESC                 0
PARKS_NM             109018
PATROL_BORO               0
PD_CD                     0
PD_DESC                   0
PREM_TYP_DESC           376
RPT_DT                    0
STATION_NAME         106348
SUSP_AGE_GROUP        28109
SUSP_RACE             28109
SUSP_SEX              28109
TRANSIT_DISTRICT     106348
VIC_AGE_GROUP             0
VIC_RACE                  0
VIC_SEX                   0
X_COORD_CD                0
Y_COORD_CD                0
Latitude                  0
Longitude                 0
Lat_Lon                   0
dtype: int64

In [227]:
data_df = data_df.dropna(subset = ['OFNS_DESC', 'X_COORD_CD', 'Y_COORD_CD', 'Latitude','Longitude','Lat_Lon'])

Selecting only few columns from the data that seem to be important and necessary for the analysis. 

1. Removing CMPLT_TO_DT and CMPLT_TO_TM columns, as they seem to be less important when considered CMPLT_FR_DT and CMPLT_FR_TM 
2. Based on the data description, HADEVELOPT and  HOUSINNG_PSA also seem to be less important. 
3. Even removing PARKS_NM because of two reasons. 1. Variable with highest missing values and assuming not all crimes take place in parks 2. PREM_TYPE_DESC has values where crime has occurred 

In [228]:
df = data_df[['CMPLNT_NUM', 'ADDR_PCT_CD', 'BORO_NM', 'CMPLNT_FR_DT', 'CMPLNT_FR_TM', 'CRM_ATPT_CPTD_CD', 'JURISDICTION_CODE', 'JURIS_DESC','KY_CD','LAW_CAT_CD', 'OFNS_DESC', 
'PD_CD','PD_DESC','PREM_TYP_DESC','RPT_DT','STATION_NAME','SUSP_AGE_GROUP','SUSP_RACE','SUSP_SEX', 'VIC_AGE_GROUP',
       'VIC_RACE', 'VIC_SEX', 'X_COORD_CD','Y_COORD_CD','Latitude','Longitude','Lat_Lon']]

In [229]:
df.head()

CMPLNT_NUM  ADDR_PCT_CD        BORO_NM CMPLNT_FR_DT CMPLNT_FR_TM  \
0   831526991           67       BROOKLYN   03/31/2018     23:30:00   
1   496736340          120  STATEN ISLAND   03/31/2018     23:20:00   
2   399593692          105         QUEENS   03/31/2018     23:15:00   
3   282376455          110         QUEENS   03/31/2018     23:10:00   
4   201598299           78       BROOKLYN   03/31/2018     23:10:00   

  CRM_ATPT_CPTD_CD  JURISDICTION_CODE           JURIS_DESC  KY_CD  \
0        COMPLETED                0.0     N.Y. POLICE DEPT    236   
1        COMPLETED                2.0  N.Y. HOUSING POLICE    104   
2        COMPLETED                0.0     N.Y. POLICE DEPT    361   
3        COMPLETED                0.0     N.Y. POLICE DEPT    105   
4        COMPLETED                2.0  N.Y. HOUSING POLICE    578   

    LAW_CAT_CD  ...       SUSP_RACE  SUSP_SEX VIC_AGE_GROUP        VIC_RACE  \
0  MISDEMEANOR  ...             NaN       NaN       UNKNOWN         UNKNOWN   
1       FELONY  ...         UNKNOWN         M           <18           BLACK   
2  MISDEMEANOR  ...           BLACK         M         18-24           BLACK   
3       FELONY  ...  WHITE HISPANIC         M         25-44  WHITE HISPANIC   
4    VIOLATION  ...  WHITE HISPANIC         F         25-44           BLACK   

  VIC_SEX X_COORD_CD Y_COORD_CD   Latitude  Longitude  \
0       E  1003227.0   177460.0  40.653751 -73.931609   
1       F   962748.0   174174.0  40.644726 -74.077483   
2       F  1056932.0   200034.0  40.715434 -73.737816   
3       M  1014990.0   210503.0  40.744414 -73.889065   
4       F   989141.0   187733.0  40.681967 -73.982367   

                         Lat_Lon  
0  (40.653751263, -73.931609227)  
1  (40.644726131, -74.077483159)  
2  (40.715433534, -73.737816163)  
3  (40.744413503, -73.889065158)  
4  (40.681967344, -73.982367054)  

[5 rows x 27 columns]

In [230]:
df.isna().sum()

CMPLNT_NUM                0
ADDR_PCT_CD               0
BORO_NM                   0
CMPLNT_FR_DT              0
CMPLNT_FR_TM              0
CRM_ATPT_CPTD_CD          0
JURISDICTION_CODE         0
JURIS_DESC                0
KY_CD                     0
LAW_CAT_CD                0
OFNS_DESC                 0
PD_CD                     0
PD_DESC                   0
PREM_TYP_DESC           376
RPT_DT                    0
STATION_NAME         106348
SUSP_AGE_GROUP        28109
SUSP_RACE             28109
SUSP_SEX              28109
VIC_AGE_GROUP             0
VIC_RACE                  0
VIC_SEX                   0
X_COORD_CD                0
Y_COORD_CD                0
Latitude                  0
Longitude                 0
Lat_Lon                   0
dtype: int64

1. Filling NA values with mode in cases of SUSP_RACE and SUSP_AGE_GROUP. 
2. Assuming that N.Y. POLICE DEPT as a central point, filling na values of STATION_NAME with the same. 
3. SUSP_SEX is more distributed towards Male and less towards Female. To get a basic or equal distribution of values, I assumed missing values to be Female. 
4. For PREM_TYPE_DESC, assuming the most crimes happen at a similar place or type of place, I used mode of PREM_TYPE_DESC to fill missing values.

In [231]:
df['SUSP_RACE'].fillna(df['SUSP_RACE'].mode()[0], inplace=True)
df['SUSP_AGE_GROUP'].fillna(df['SUSP_AGE_GROUP'].mode()[0], inplace=True)
df['STATION_NAME'].fillna('N.Y. POLICE DEPT', inplace=True)
df['SUSP_SEX'].fillna('F', inplace=True)
df['PREM_TYP_DESC'].fillna(df['PREM_TYP_DESC'].mode()[0], inplace=True)

/Users/sarathchandrikak/opt/anaconda3/envs/mysqlconnect/lib/python3.6/site-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [232]:
df.isna().sum()

CMPLNT_NUM           0
ADDR_PCT_CD          0
BORO_NM              0
CMPLNT_FR_DT         0
CMPLNT_FR_TM         0
CRM_ATPT_CPTD_CD     0
JURISDICTION_CODE    0
JURIS_DESC           0
KY_CD                0
LAW_CAT_CD           0
OFNS_DESC            0
PD_CD                0
PD_DESC              0
PREM_TYP_DESC        0
RPT_DT               0
STATION_NAME         0
SUSP_AGE_GROUP       0
SUSP_RACE            0
SUSP_SEX             0
VIC_AGE_GROUP        0
VIC_RACE             0
VIC_SEX              0
X_COORD_CD           0
Y_COORD_CD           0
Latitude             0
Longitude            0
Lat_Lon              0
dtype: int64

# Data Validation

In [233]:
df['SUSP_SEX'].value_counts()

M    50864
F    43633
U    14982
Name: SUSP_SEX, dtype: int64

In [234]:
data_df['VIC_SEX'].value_counts()

F    42823
M    35336
E    15699
D    15621
Name: VIC_SEX, dtype: int64

VIC_SEX has E and D which is not mentioned anywhere in Data description. So, I am assuming them as Female. Since it might be a mistype to F because both E and D are adjacent to F on the keyboard. 

In [235]:
df['VIC_SEX'] = df['VIC_SEX'].replace(['E', 'D'],'F')

/Users/sarathchandrikak/opt/anaconda3/envs/mysqlconnect/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [236]:
data_df['VIC_AGE_GROUP'].value_counts()

25-44      36445
UNKNOWN    34025
45-64      19623
18-24       9956
<18         4949
65+         4473
-974           2
972            1
-972           1
951            1
-942           1
-958           1
-43            1
Name: VIC_AGE_GROUP, dtype: int64

Upon observing the data selecting few age groups from the data. Following are the selected age groups for victims

In [237]:
correct_age = ['25-44', 'UNKNOWN', '45-64', '18-24', '<18', '64+']
df = df[df['VIC_AGE_GROUP'].isin(correct_age)]

In [238]:
age_group_count = df['VIC_AGE_GROUP'].value_counts()
age_group_count

25-44      36445
UNKNOWN    34025
45-64      19623
18-24       9956
<18         4949
Name: VIC_AGE_GROUP, dtype: int64

In [239]:
age_group_percent = df['VIC_AGE_GROUP'].value_counts(normalize=True)
age_group_percent

25-44      0.347102
UNKNOWN    0.324054
45-64      0.186889
18-24      0.094821
<18        0.047134
Name: VIC_AGE_GROUP, dtype: float64

In [240]:
df['SUSP_AGE_GROUP'].value_counts()

25-44      54222
UNKNOWN    26128
18-24      10439
45-64      10111
<18         3353
65+          732
2018           3
928            2
1018           2
952            1
954            1
920            1
-2             1
1967           1
938            1
Name: SUSP_AGE_GROUP, dtype: int64

Similar way selecting valid age group for suspects

In [241]:
correct_age_susp = ['25-44', 'UNKNOWN', '18-24', '45-64', '<18', '65+']

In [242]:
df = df[df['SUSP_AGE_GROUP'].isin(correct_age_susp)]

In [243]:
df['SUSP_AGE_GROUP'].value_counts()

25-44      54222
UNKNOWN    26128
18-24      10439
45-64      10111
<18         3353
65+          732
Name: SUSP_AGE_GROUP, dtype: int64

In [244]:
df.dtypes

CMPLNT_NUM             int64
ADDR_PCT_CD            int64
BORO_NM               object
CMPLNT_FR_DT          object
CMPLNT_FR_TM          object
CRM_ATPT_CPTD_CD      object
JURISDICTION_CODE    float64
JURIS_DESC            object
KY_CD                  int64
LAW_CAT_CD            object
OFNS_DESC             object
PD_CD                float64
PD_DESC               object
PREM_TYP_DESC         object
RPT_DT                object
STATION_NAME          object
SUSP_AGE_GROUP        object
SUSP_RACE             object
SUSP_SEX              object
VIC_AGE_GROUP         object
VIC_RACE              object
VIC_SEX               object
X_COORD_CD           float64
Y_COORD_CD           float64
Latitude             float64
Longitude            float64
Lat_Lon               object
dtype: object

Validating date values and cleaning the anomalies 

In [245]:
df['CMPLNT_FR_DT_TM'] = df['CMPLNT_FR_DT'] + ' ' + df['CMPLNT_FR_TM'] 

In [246]:
print(df['CMPLNT_FR_DT_TM'].head())

0    03/31/2018 23:30:00
1    03/31/2018 23:20:00
2    03/31/2018 23:15:00
3    03/31/2018 23:10:00
4    03/31/2018 23:10:00
Name: CMPLNT_FR_DT_TM, dtype: object


In [247]:
from datetime import datetime
datetime_list = []
for d in df['CMPLNT_FR_DT_TM']:
    datetime_list.append(datetime.strptime(d, '%m/%d/%Y %H:%M:%S'))

Getting all the year values that are less than 1900

In [248]:
for i in datetime_list:
    if i.year < 1900:
        print(i.year)

1018
1028
1018
1028
1018
1018
1016
1018


In [249]:
for i in range(len(datetime_list)):
    if datetime_list[i].year < 1900:
        x = datetime_list[i].year + 1000
        updated = datetime(x, datetime_list[i].month, datetime_list[i].day, 0, 0, 0)
        datetime_list[i] = updated        

Checking if there are any more values less than 1900

In [250]:
for i in datetime_list:
    if i.year < 1900:
        print(i.year)

In [251]:
df['CMPLNT_FR_DT_TM'] = datetime_list
df['CMPLNT_FR_DT_TM'] = pd.to_datetime(df['CMPLNT_FR_DT_TM'])

In [255]:
print(df.dtypes)

CMPLNT_NUM                    int64
ADDR_PCT_CD                   int64
BORO_NM                      object
CMPLNT_FR_DT                 object
CMPLNT_FR_TM                 object
CRM_ATPT_CPTD_CD             object
JURISDICTION_CODE           float64
JURIS_DESC                   object
KY_CD                         int64
LAW_CAT_CD                   object
OFNS_DESC                    object
PD_CD                       float64
PD_DESC                      object
PREM_TYP_DESC                object
RPT_DT                       object
STATION_NAME                 object
SUSP_AGE_GROUP               object
SUSP_RACE                    object
SUSP_SEX                     object
VIC_AGE_GROUP                object
VIC_RACE                     object
VIC_SEX                      object
X_COORD_CD                  float64
Y_COORD_CD                  float64
Latitude                    float64
Longitude                   float64
Lat_Lon                      object
CMPLNT_FR_DT_TM      datetim

In [253]:
df = df[df['CMPLNT_FR_DT_TM'].dt.year < pd.datetime.now().year]

/Users/sarathchandrikak/opt/anaconda3/envs/mysqlconnect/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime instead.
  """Entry point for launching an IPython kernel.


In [254]:
print(max(df['CMPLNT_FR_DT_TM']))

2018-03-31 23:30:00


In [214]:
df.head()

CMPLNT_NUM  ADDR_PCT_CD        BORO_NM CMPLNT_FR_DT CMPLNT_FR_TM  \
0   831526991           67       BROOKLYN   03/31/2018     23:30:00   
1   496736340          120  STATEN ISLAND   03/31/2018     23:20:00   
2   399593692          105         QUEENS   03/31/2018     23:15:00   
3   282376455          110         QUEENS   03/31/2018     23:10:00   
4   201598299           78       BROOKLYN   03/31/2018     23:10:00   

  CRM_ATPT_CPTD_CD  JURISDICTION_CODE           JURIS_DESC  KY_CD  \
0        COMPLETED                0.0     N.Y. POLICE DEPT    236   
1        COMPLETED                2.0  N.Y. HOUSING POLICE    104   
2        COMPLETED                0.0     N.Y. POLICE DEPT    361   
3        COMPLETED                0.0     N.Y. POLICE DEPT    105   
4        COMPLETED                2.0  N.Y. HOUSING POLICE    578   

    LAW_CAT_CD  ... SUSP_SEX  VIC_AGE_GROUP        VIC_RACE VIC_SEX  \
0  MISDEMEANOR  ...        F        UNKNOWN         UNKNOWN       F   
1       FELONY  ...        M            <18           BLACK       F   
2  MISDEMEANOR  ...        M          18-24           BLACK       F   
3       FELONY  ...        M          25-44  WHITE HISPANIC       M   
4    VIOLATION  ...        F          25-44           BLACK       F   

  X_COORD_CD Y_COORD_CD   Latitude  Longitude                        Lat_Lon  \
0  1003227.0   177460.0  40.653751 -73.931609  (40.653751263, -73.931609227)   
1   962748.0   174174.0  40.644726 -74.077483  (40.644726131, -74.077483159)   
2  1056932.0   200034.0  40.715434 -73.737816  (40.715433534, -73.737816163)   
3  1014990.0   210503.0  40.744414 -73.889065  (40.744413503, -73.889065158)   
4   989141.0   187733.0  40.681967 -73.982367  (40.681967344, -73.982367054)   

      CMPLNT_FR_DT_TM  
0 2018-03-31 23:30:00  
1 2018-03-31 23:20:00  
2 2018-03-31 23:15:00  
3 2018-03-31 23:10:00  
4 2018-03-31 23:10:00  

[5 rows x 28 columns]

In [256]:
df.drop(['CMPLNT_FR_DT', 'CMPLNT_FR_TM'], axis=1, inplace=True)

In [257]:
df.columns

Index(['CMPLNT_NUM', 'ADDR_PCT_CD', 'BORO_NM', 'CRM_ATPT_CPTD_CD',
       'JURISDICTION_CODE', 'JURIS_DESC', 'KY_CD', 'LAW_CAT_CD', 'OFNS_DESC',
       'PD_CD', 'PD_DESC', 'PREM_TYP_DESC', 'RPT_DT', 'STATION_NAME',
       'SUSP_AGE_GROUP', 'SUSP_RACE', 'SUSP_SEX', 'VIC_AGE_GROUP', 'VIC_RACE',
       'VIC_SEX', 'X_COORD_CD', 'Y_COORD_CD', 'Latitude', 'Longitude',
       'Lat_Lon', 'CMPLNT_FR_DT_TM'],
      dtype='object')

Grouping timings as morning, afternoon, evening and night

In [263]:
df['CMPLNT_FR_DT_TM_SESSION'] = df['CMPLNT_FR_DT_TM']

In [264]:
df.dtypes

CMPLNT_NUM                          int64
ADDR_PCT_CD                         int64
BORO_NM                            object
CRM_ATPT_CPTD_CD                   object
JURISDICTION_CODE                 float64
JURIS_DESC                         object
KY_CD                               int64
LAW_CAT_CD                         object
OFNS_DESC                          object
PD_CD                             float64
PD_DESC                            object
PREM_TYP_DESC                      object
RPT_DT                             object
STATION_NAME                       object
SUSP_AGE_GROUP                     object
SUSP_RACE                          object
SUSP_SEX                           object
VIC_AGE_GROUP                      object
VIC_RACE                           object
VIC_SEX                            object
X_COORD_CD                        float64
Y_COORD_CD                        float64
Latitude                          float64
Longitude                         

In [265]:
df['CMPLNT_FR_DT_TM_SESSION'] = (df['CMPLNT_FR_DT_TM_SESSION'].dt.hour % 24 + 4) // 4
df['CMPLNT_FR_DT_TM_SESSION'].replace({1: 'Late Night',
                      2: 'Early Morning',
                      3: 'Morning',
                      4: 'Noon',
                      5: 'Evening',
                      6: 'Night'}, inplace=True)

In [266]:
print(df['CMPLNT_FR_DT_TM_SESSION'])

0              Night
1              Night
2              Night
3              Night
4              Night
             ...    
109538          Noon
109539         Night
109540          Noon
109541       Morning
109542    Late Night
Name: CMPLNT_FR_DT_TM_SESSION, Length: 104983, dtype: object


In [109]:
df.to_csv('nyc_crimes_modified.csv')

# Descriptive Analysis

1. Try looking at precincts? (police district areas)
2. CRM_ATPT_CPTD_CD...look at the data dictionary and see if this is interesting
3. Weekend vs Weekday (would take some data cleaning)
4. LAW_CAT_CD? 
5. PARKS_NM?
6. STATION_NAME?

In [307]:
data_df.columns

Index(['CMPLNT_NUM', 'ADDR_PCT_CD', 'BORO_NM', 'CMPLNT_FR_DT', 'CMPLNT_FR_TM',
       'CMPLNT_TO_DT', 'CMPLNT_TO_TM', 'CRM_ATPT_CPTD_CD', 'HADEVELOPT',
       'HOUSING_PSA', 'JURISDICTION_CODE', 'JURIS_DESC', 'KY_CD', 'LAW_CAT_CD',
       'LOC_OF_OCCUR_DESC', 'OFNS_DESC', 'PARKS_NM', 'PATROL_BORO', 'PD_CD',
       'PD_DESC', 'PREM_TYP_DESC', 'RPT_DT', 'STATION_NAME', 'SUSP_AGE_GROUP',
       'SUSP_RACE', 'SUSP_SEX', 'TRANSIT_DISTRICT', 'VIC_AGE_GROUP',
       'VIC_RACE', 'VIC_SEX', 'X_COORD_CD', 'Y_COORD_CD', 'Latitude',
       'Longitude', 'Lat_Lon'],
      dtype='object')

In [310]:
print(len(df['BORO_NM'].unique()))

5


In [311]:
df['BORO_NM'].value_counts()

BROOKLYN         30957
MANHATTAN        26331
BRONX            23518
QUEENS           19572
STATEN ISLAND     4605
Name: BORO_NM, dtype: int64

A total of 5 different Police districts with BROOKLYN recording the highest number of crimes and STATEN ISLAND recording the lowest number of crimes

In [267]:
df['CRM_ATPT_CPTD_CD'].value_counts()

COMPLETED    103196
ATTEMPTED      1787
Name: CRM_ATPT_CPTD_CD, dtype: int64

Value counts function indicated that there are in total of around 100k completed incidents and 1800 of attempted incidents. From this it can be stated that most of the crimes are successfully completed (which is pretty sad)

To compare weekday vs weekend

In [288]:
df['DAY_TYPE'] = df['CMPLNT_FR_DT_TM']
df['DAY_TYPE'] = df['DAY_TYPE'].dt.day_name()
df['DAY_TYPE'].replace({'Saturday': 'weekend',
                        'Sunday' : 'weekend',
                        'Monday' : 'weekday',
                        'Tuesday' : 'weekday',
                        'Wednesday' : 'weekday', 
                        'Thursday' : 'weekday', 
                        'Friday': 'weekday'},inplace=True)

In [289]:
print(df['DAY_TYPE'])

0         weekend
1         weekend
2         weekend
3         weekend
4         weekend
           ...   
109538    weekday
109539    weekday
109540    weekday
109541    weekday
109542    weekday
Name: DAY_TYPE, Length: 104983, dtype: object


In [290]:
df['DAY_TYPE'].value_counts()

weekday    76694
weekend    28289
Name: DAY_TYPE, dtype: int64

From the observations, it is seen that more number of crimes occured during weekdays rather than weekends. 

In [291]:
df['LAW_CAT_CD'].value_counts()

MISDEMEANOR    58885
FELONY         30705
VIOLATION      15393
Name: LAW_CAT_CD, dtype: int64

Based on the data it is seen that Misdemeanor has more count Level followed by felony and violation.

In [296]:
data_df['PARKS_NM'].isna().sum()

109018

In [295]:
data_df['PARKS_NM'].value_counts()

FLUSHING MEADOWS CORONA PARK       32
CENTRAL PARK                       24
WILLIAMSBRIDGE OVAL                17
RIVERSIDE PARK                     15
UNION SQUARE PARK                  13
                                   ..
CASTLE HILL LITTLE LEAGUE FIELD     1
WINGATE PARK                        1
BRONX RIVER PARKWAY                 1
MCCAFFREY PLAYGROUND                1
PLAYGROUND 286                      1
Name: PARKS_NM, Length: 201, dtype: int64

Since there are many null values in PARKS_NM column, I removed the column in my subset dataframe. But as per the exploration, Flushing Meadows corona park has highest number of crimes reported and many gave just reported 1 crime. 

In [304]:
n = len(pd.unique(df['STATION_NAME']))
print(n)

328


In [305]:
df['STATION_NAME'].value_counts()

N.Y. POLICE DEPT                  101898
125 STREET                           131
42 ST.-PORT AUTHORITY BUS TERM        98
14 STREET                             84
59 ST.-COLUMBUS CIRCLE                73
                                   ...  
36 AVENUE                              1
CHRISTOPHER ST.-SHERIDAN SQ.           1
DISTRICT 20 OFFICE                     1
BUHRE AVENUE                           1
HEWES STREET                           1
Name: STATION_NAME, Length: 328, dtype: int64

Total crimes have been reported in 328 different station names and out of then N.Y. POLICE DEPT has highest number of crimes reported.